Improvements to user experience

Aurelien30000 opened the pull request:
---

### Archive artifacts on Github, following FAWE changes.

---

### Improve and fix #getLastBlock() method

- Previously returned the wrong block when performing actions far from the ground (null targetBlock in Bukkit interaction event).

---

### Improve some command executors

- Remove duplicate parts of code
- Improve block tab-completion permitting whether ``minecraft:XXXX`` or ``XXXX``

---

### Major improvements to commands:
##
- Add a tab-completion method to ``Brush`` interface, allowing each brush to complete it own parameters.
- Add javadocs to ``Brush`` interface.
- Don't add an empty argument to commands parameters. _This was added for an unknown reason and is more likely unneeded now._
- Improve brush info, propagating a sane common _format_.
- Avoid using ``String#replace`` to substring a given amount of chars.
##
- Add a tab-completion handler to brushes that accept parameters, such as biomes, block types or precise arguments.
- Thoses handlers have been added to improve users' usage and understanding. However, they might be removed in the future to hand hover to a hypothetical switch to a modern command framework (as @NotMyFault suggested).
- Each brush can override the default handler. Even though a brush doesn't implement or need tab-completion, ``info`` parameter is always suggested as the first one, among others if given obviously.
- Rewrite ``EntityBrush`` to merge WE and Bukkit APIs. This brings protection against ability to add entities anywhere, thanks to this brush.
- Close IntellectualSites#49
- Use real subcommands instead of parsed parameters such as ``y15``.
##
- Minor refactor due to previous changes (useless parameters, typos).
- Update to latest FAWE changes, remove FAVS workarounds about _Tree_ brushes.

---

```
This PR will certainly be the last one to bring _so many changes_. Since I'll have to get back to studies soon,
it will be difficult to spend so much time working on FAVS, but I intend to continue contributing step by step!

To sum up, this PR doesn't bring many technical changes and aims at improving user experience.
```

Aurelien30000 committed:
Archive artifacts and bump FAWE version

Aurelien30000 committed:
Use real version to upload artifacts

Aurelien30000 committed:
Revert "bump FAWE version"

NotMyFault committed:
2.1.3

Aurelien30000 committed:
Improve artifacts uploading.

Aurelien30000 committed:
Improve and fix ``#getLastBlock()`` method
- Previously returned the wrong block when performing actions far from the ground (null targetBlock in Bukkit interaction event).

Aurelien30000 committed:
Improve some command executors
- Remove duplicate parts of code
- Improve block tab-completion permitting whether ``minecraft:XXXX`` or ``XXXX``

Aurelien30000 committed:
Major improvements to commands:
- Add a tab-completion method to ``Brush`` interface, allowing each brush to complete it own parameters.
- Add javadocs to ``Brush`` interface.
- Don't add an empty argument to commands parameters. _This was added for an unknown reason and is more likely unneeded now._
- Improve brush info, propagating a sane common _format_.
- Avoid using ``String#replace`` to substring a given amount of chars.

- Add a tab-completion handler to brushes that accept parameters, such as biomes, block types or precise arguments.
- Thoses handlers have been added to improve users' usage and understanding. However, they might be removed in the future to hand hover to a hypothetical switch to a modern command framework (as @NotMyFault suggested).
- Each brush can override the default handler. Even though a brush doesn't implement or need tab-completion, ``info`` parameter is always suggested as the first one, among others if given obviously.
- Rewrite ``EntityBrush`` to merge WE and Bukkit APIs. This brings protection against ability to add entities anywhere, thanks to this brush.
- Close https://github.com/IntellectualSites/FastAsyncVoxelSniper/issues/49

- Minor refactor due to previous changes (useless parameters, typos).

Aurelien30000 committed:
Fix build

NotMyFault commented on .github/workflows/build.yml:
```diff
@@ -8,10 +8,24 @@ jobs:
+        uses : actions/upload-artifact@v2
+        with :
+          name : FastAsyncVoxelSniper-2.1.3-${{ env.SHA }}
+          path : build/libs/FastAsyncVoxelSniper-*-${{ env.SHA }}.jar
```
I'd rather not hardcode the version string here. Either use regex, cache the version from `./gradlew properties` or use something else suitable that doesn't rely on 1 set version.

NotMyFault commented on build.gradle.kts:
```diff
@@ -45,7 +45,7 @@ configurations.all {
 }
 
-var rootVersion by extra("2.1.1")
+var rootVersion by extra("2.1.3")
```
```suggestion
var rootVersion by extra("2.2.0")
```

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/BiomeBrush.java:
```diff
@@ -2,50 +2,54 @@
-    private Biome selectedBiome = Biome.PLAINS;
+    private static final List<String> BIOMES = BiomeTypes.values().stream()
+            .map(BiomeType::getId)
+            .collect(Collectors.toList());
```
Register biomes from NMS to avoid the issue https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/1200 addresses.

Aurelien30000 replied:
Since FAVS depends on FAWE, is it really necessary to register biomes twice?

NotMyFault replied:
No, FAVS loads after FAWE so it's safe to assume necessary hooks are set up by then.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/BiomeBrush.java:
```diff
@@ -2,50 +2,54 @@
+
+    @Override
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
```
This will cut tab completion after `m` if you attempt to tab complete biomes with minecraft:. In fact, the default tab completion doesn't need to display `minecraft:` at all. It will work without nevertheless, see how (FA)WE handles tab completion here for example.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/CometBrush.java:
```diff
@@ -23,7 +26,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+                messenger.sendMessage("/b com balls [big|small]  -- Sets your ball size.");
             }
             if (parameter.equalsIgnoreCase("balls")) {
                 if (index + 1 >= parameters.length) {
```
Next line will cause an uncaught AIOOBE if no args are used. `/b com balls`

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/CometBrush.java:
```diff
@@ -43,6 +46,19 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+        }
+        return super.handleCompletions(parameters, snipe);
+    }
+
```
I'll put this here but it may affect other tab completions too.
![](https://i.imgur.com/TGiDz4d.png)
This format is out of place and may fails on some users.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/CopyPastaBrush.java:
```diff
@@ -50,6 +51,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         }
```
Can only comment here but
```java
 if (Stream.of("90", "180", "270", "0")
                .anyMatch(parameter::equalsIgnoreCase)) {
```
Maybe drop a message if the player tries to use another number that the pivot angle doesn't support?

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ErodeBrush.java:
```diff
@@ -120,6 +121,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("f", "e", "F", "E"), parameter, 0);
```
What are those letters? Is there some form of explanation? `/b eb info` returns the info for `/b bb`

Aurelien30000 replied:
_``eb`` brush is ErodeBlend ``e`` is erode_

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/FlatOceanBrush.java:
```diff
@@ -23,18 +26,18 @@ public void handleCommand(String[] parameters, Snipe snipe) {
-                messenger.sendMessage(ChatColor.GREEN + "yo[number] to set the Level to which the water will rise.");
-                messenger.sendMessage(ChatColor.GREEN + "yl[number] to set the Level to which the ocean floor will rise.");
+                messenger.sendMessage(ChatColor.GREEN + "/b fo yo[number] to set the Level to which the water will rise.");
+                messenger.sendMessage(ChatColor.GREEN + "/b fo yl[number] to set the Level to which the ocean floor will rise.");
```
Lowercase "Level"

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/FlatOceanBrush.java:
```diff
@@ -23,18 +26,18 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 messenger.sendMessage(ChatColor.GREEN + "Water Level set to " + this.waterLevel);
             } else if (parameter.startsWith("yl")) {
-                int newFloorLevel = Integer.parseInt(parameter.replace("yl", ""));
+                int newFloorLevel = Integer.parseInt(parameter.substring(2));
```
Optimally that'd be revised. The current input needs to be done like `/b fo yl5`. We should follow the standard and put a space there and parse it like `/b fo yl 5`.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/FlatOceanBrush.java:
```diff
@@ -48,6 +51,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("yo", "yl"), parameter, 0);
```
See my former comment, the tab completion would need to be adjusted if doing so.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/GenerateTreeBrush.java:
```diff
@@ -59,104 +60,104 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                     messenger.sendMessage(ChatColor.BLUE + "Leaf Type set to " + this.leafType.getRichName().toString());
                 } else if (parameter.startsWith("wt")) { // Wood Type
-                    BlockType woodType = BlockTypes.get((parameter.replace("wt", "") + "_log")
+                    BlockType woodType = BlockTypes.get((parameter.substring(2) + "_log")
```
See my comment at the FlatOcean brush. Parsing ints without a space and clarification brackets is very legacy. I'll put this here but it may affect other classes too needing some revision.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/HeatRayBrush.java:
```diff
@@ -64,18 +66,27 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 messenger.sendMessage(ChatColor.GREEN + "Amplitude: " + this.amplitude);
             } else if (parameter.startsWith("freq")) {
-                this.frequency = Double.parseDouble(parameter.replace("freq", ""));
+                this.frequency = Double.parseDouble(parameter.substring(4));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/JockeyBrush.java:
```diff
@@ -57,6 +58,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("-i:", "-po:", "-s:"), parameter, 0);
```
The info parameter is appended but doesn't provide any info what the args do. Also why do they use a dash, for sake of consistency that shouldn't be necessary.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/MoveBrush.java:
```diff
@@ -65,6 +66,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         }
     }
 
+    @Override
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/PullBrush.java:
```diff
@@ -23,8 +23,8 @@
+        Double bubble = NumericParser.parseDouble(parameters[0]);
         if (pinch == null || bubble == null) {
             SnipeMessenger messenger = snipe.createMessenger();
             messenger.sendMessage(ChatColor.RED + "Invalid brush parameters!");
```
Using the `info` parameter sends this message.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RegenerateChunkBrush.java:
```diff
@@ -18,35 +17,57 @@
+                String parameter = parameters[1];
+                return super.sortCompletions(Stream.concat(
+                        Stream.of(DEFAULT_BIOME),
+                        BIOMES.stream()
```
See comment with previous NMS

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RulerBrush.java:
```diff
@@ -30,13 +33,13 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 return;
             } else if (parameter.charAt(0) == 'x') {
-                this.offsetX = Integer.parseInt(parameter.replace("x", ""));
+                this.offsetX = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ScannerBrush.java:
```diff
@@ -46,6 +49,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         }
     }
 
+    @Override
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -54,15 +56,16 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+                            .message(ChatColor.GREEN + "/b sio -clearall " + ChatColor.BLUE + "-- Clears the line buffer and sets all lines back to enabled. (Alias: -ca)")
+                            .message(ChatColor.GREEN + "/b sio -multiple [on|off] " + ChatColor.BLUE + "-- Enables or disables ranged mode. (Alias: -m) (see Wiki for more information)")
+                            .message(ChatColor.GREEN + "/b sio -save (name) " + ChatColor.BLUE + "-- Save you buffer to a file named [name]. (Alias: -s)")
+                            .message(ChatColor.GREEN + "/b sio -open (name) " + ChatColor.BLUE + "-- Loads a buffer from a file named [name]. (Alias: -o)")
```
See FlatOceanBrush comment and don't make the input black, that's hard to read ingame.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SnowConeBrush.java:
```diff
@@ -12,15 +12,6 @@
 
 public class SnowConeBrush extends AbstractBrush {
 
-    @Override
```
The info parameter doesn't work here.

Aurelien30000 replied:
Expected
![image](https://user-images.githubusercontent.com/43724816/128517046-882b9ef0-7959-4615-a447-cf1568bcca41.png)

NotMyFault replied:
Yes, but tab completion suggests it nevertheless.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/TreeSnipeBrush.java:
```diff
@@ -10,36 +10,53 @@
+                this.treeType = treeType;
+                messenger.sendMessage(ChatColor.GOLD + "Currently selected tree type: " + ChatColor.DARK_GREEN + this.treeType.getName());
+            } else {
                 messenger.sendMessage(ChatColor.LIGHT_PURPLE + "No such tree type.");
```
That message is thrown if no input is provided, however, the default is an oak tree. Also, any input doesn't do anything?

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/canyon/CanyonBrush.java:
```diff
@@ -18,12 +18,12 @@
         }
         if (!firstParameter.isEmpty() && firstParameter.charAt(0) == 'y') {
-            int y = Integer.parseInt(firstParameter.replace("y", ""));
+            int y = Integer.parseInt(firstParameter.substring(1));
```
See FlatOceanBrush comment and maybe leave a note if y is outside of the hardcoded range.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/entity/EntityBrush.java:
```diff
@@ -4,43 +4,54 @@
-                    .map(currentEntity -> ChatColor.AQUA + " | " + ChatColor.DARK_GREEN + currentEntity.getName())
-                    .collect(Collectors.joining("", "", ChatColor.AQUA + " |"));
-            messenger.sendMessage(names);
+        if (parameters[0].equalsIgnoreCase("info")) {
```
Might be worth to have a `list` argument like for the tree generator listing down all possible entities in the chat.

Aurelien30000 replied:
Isn't tab-completion enough? VS previously sent a list, a huge list in the chat, kinda annoying imo.
_Brushes no longer print all possible types._

NotMyFault replied:
Having a list allows you to view everything at a glance. This doesn't need to be an _or_. We can have a list args and tab completion.

Aurelien30000 replied:
Sure :D

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/entity/EntityRemovalBrush.java:
```diff
@@ -29,7 +30,10 @@ public EntityRemovalBrush() {
     public void handleCommand(String[] parameters, Snipe snipe) {
         SnipeMessenger messenger = snipe.createMessenger();
         for (String currentParam : parameters) {
-            if (!currentParam.isEmpty() && (currentParam.charAt(0) == '+' || currentParam.charAt(0) == '-')) {
```
Not validating the input is quite dangling. Looking at the code `/b entityremoval +#` would display an empty space in the chat e.g. or allow any kind of typo to screw up.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/BlobBrush.java:
```diff
@@ -26,11 +26,11 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             }
             if (parameter.charAt(0) == 'g') {
-                int temp = Integer.parseInt(parameter.replace("g", ""));
+                int temp = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment, tab completion is missing.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/CylinderBrush.java:
```diff
@@ -22,7 +25,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
-                messenger.sendMessage(ChatColor.AQUA + "/b c h[number] -- set the cylinder v.voxelHeight.  Default is 1.");
+                messenger.sendMessage(ChatColor.AQUA + "/b c h[number] -- set the cylinder v.voxelHeight. Default is 1.");
                 messenger.sendMessage(ChatColor.DARK_AQUA + "/b c true -- will use a true circle algorithm instead of the skinnier version with classic sniper nubs. /b b false will switch back. (false is default)");
                 messenger.sendMessage(ChatColor.DARK_BLUE + "/b c c[number] -- set the origin of the cylinder compared to the target block. Positive numbers will move the cylinder upward, negative will move it downward.");
```
See FlatOceanBrush comment & dark blue is a horrible color to display anything ingame. Putting this here for the entire project, replace unreadable colors like black or dark blue in spotted.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/EllipseBrush.java:
```diff
@@ -30,13 +33,13 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         for (String parameter : parameters) {
```
See FlatOceanBrush comment

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/EllipsoidBrush.java:
```diff
@@ -23,18 +26,18 @@ public void handleCommand(String[] parameters, Snipe snipe) {
-                    messenger.sendMessage(ChatColor.AQUA + "x[n]: Set X radius to n");
-                    messenger.sendMessage(ChatColor.AQUA + "y[n]: Set Y radius to n");
-                    messenger.sendMessage(ChatColor.AQUA + "z[n]: Set Z radius to n");
+                    messenger.sendMessage(ChatColor.AQUA + "/b elo x[n]: Set X radius to n");
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/FillDownBrush.java:
```diff
@@ -56,6 +59,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("true", "false", "some", "all", "-e"), parameter, 0);
```
Move -e to a regular sub command.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/JaggedLineBrush.java:
```diff
@@ -67,6 +69,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         }
     }
 
+    @Override
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/LineBrush.java:
```diff
@@ -22,7 +22,7 @@
-        if (parameters[1].equalsIgnoreCase("info")) {
+        if (parameters[0].equalsIgnoreCase("info")) {
             SnipeMessenger messenger = snipe.createMessenger();
             messenger.sendMessage(ChatColor.GOLD + "Line Brush instructions: Right click first point with the arrow. Right click with powder to draw a line to set the second point.");
```
Putting this here for the entire project: Gunpower > powder. Especially in days of powdersnow and concrete powder. Replace on sight.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java:
```diff
@@ -26,13 +29,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             if (parameter.charAt(0) == 'd') {
                 try {
-                    this.depth = Integer.parseInt(parameter.replace("d", ""));
+                    this.depth = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/RingBrush.java:
```diff
@@ -31,7 +34,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             } else if (parameter.startsWith("ir")) {
                 try {
-                    this.innerSize = Double.parseDouble(parameter.replace("ir", ""));
+                    this.innerSize = Double.parseDouble(parameter.substring(2));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/SplineBrush.java:
```diff
@@ -73,6 +74,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
         }
     }
 
+    @Override
```
Putting this here, but using `ren` would produce an AIOOBE if used before setting control- and endpoints.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java:
```diff
@@ -24,16 +27,17 @@ public final void handleCommand(String[] parameters, Snipe snipe) {
             }
             if (parameter.charAt(0) == 'd') {
-                this.depth = Integer.parseInt(parameter.replace("d", ""));
+                this.depth = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java:
```diff
@@ -24,16 +27,17 @@ public final void handleCommand(String[] parameters, Snipe snipe) {
+                this.depth = Integer.parseInt(parameter.substring(1));
                 messenger.sendMessage(ChatColor.AQUA + "Depth set to " + this.depth);
             } else if (parameter.startsWith("all")) {
                 this.allBlocks = true;
```
Putting this here for all following args, missing space between message and variable.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterBallBrush.java:
```diff
@@ -41,23 +43,23 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 return;
             } else if (parameter.charAt(0) == 's') {
-                double temp = Integer.parseInt(parameter.replace("s", ""));
+                double temp = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterDiscBrush.java:
```diff
@@ -41,23 +43,23 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 return;
             } else if (parameter.charAt(0) == 's') {
-                double temp = Integer.parseInt(parameter.replace("s", ""));
+                double temp = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java:
```diff
@@ -44,15 +46,17 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                     return;
                 } else if (parameter.charAt(0) == 'd') {
-                    this.depth = Integer.parseInt(parameter.replace("d", ""));
+                    this.depth = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java:
```diff
@@ -107,6 +111,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("d", "all", "some", "s", "g", "r", "randh", "yoff"), parameter, 0);
```
Starting the args with `s` or `r` would complete some or randh. Individual letters aren't treated properly.

Aurelien30000 replied:
I don't know what to do here 😓 the handler is only suggesting an alternative. Should nothing be suggested if an existing valid argument is given?

NotMyFault replied:
Are `s` and `r` aliases some & randh?

Aurelien30000 replied:
No

NotMyFault replied:
Oic,, the existing behavior must be sufficient then without changing existing 1 letter args. Tho that could be better for sake intelligibility.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterVoxelBrush.java:
```diff
@@ -41,23 +43,23 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 return;
             } else if (parameter.charAt(0) == 's') {
-                double temp = Integer.parseInt(parameter.replace("s", ""));
+                double temp = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterVoxelDiscBrush.java:
```diff
@@ -41,23 +43,23 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                 return;
             } else if (parameter.charAt(0) == 's') {
-                double temp = Integer.parseInt(parameter.replace("s", ""));
+                double temp = Integer.parseInt(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/redstone/SetRedstoneFlipBrush.java:
```diff
@@ -28,7 +29,8 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             if (parameter.equalsIgnoreCase("info")) {
                 messenger.sendMessage(ChatColor.GOLD + "Set Repeater Flip Parameters:");
-                messenger.sendMessage(ChatColor.AQUA + "/b setrf <direction> -- valid direction inputs are(n,s,e,world), Set the direction that you wish to flip your repeaters, defaults to north/south.");
+                messenger.sendMessage(ChatColor.AQUA + "/b setrf <direction> -- valid direction inputs are(n, s, ns, e, world, " +
```
Missing space.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/redstone/SetRedstoneFlipBrush.java:
```diff
@@ -45,6 +47,15 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+    public List<String> handleCompletions(String[] parameters, Snipe snipe) {
+        if (parameters.length == 1) {
+            String parameter = parameters[0];
+            return super.sortCompletions(Stream.of("n", "s", "ns", "e", "world", "ew"), parameter, 0);
```
Why would `world` be treated as east/west? I'd remove it, that is no cardinal direction.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation2DBrush.java:
```diff
@@ -20,7 +20,7 @@
     @Override
     public void handleCommand(String[] parameters, Snipe snipe) {
-        this.angle = Math.toRadians(Double.parseDouble(parameters[1]));
+        this.angle = Math.toRadians(Double.parseDouble(parameters[0]));
```
Putting this here, using the info args will produce an AIOOBE.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation2DVerticalBrush.java:
```diff
@@ -24,9 +24,9 @@
+        Double angle = NumericParser.parseDouble(parameters[0]);
         if (angle == null) {
-            messenger.sendMessage("Exception while parsing parameter: " + parameters[1]);
+            messenger.sendMessage("Exception while parsing parameter: " + parameters[0]);
```
That will cause an exception using the info args.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation3DBrush.java:
```diff
@@ -32,32 +35,49 @@ public void handleCommand(String[] parameters, Snipe snipe) {
-                if (this.seYaw < 0 || this.seYaw > 359) {
-                    messenger.sendMessage(ChatColor.RED + "Invalid brush parameters! Angles must be from 1-359");
+            } else {
+                double value = Math.toRadians(Double.parseDouble(parameter.substring(1)));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/shell/ShellVoxelBrush.java:
```diff
@@ -10,16 +10,6 @@
 
 public class ShellVoxelBrush extends AbstractBrush {
 
-    @Override
-    public void handleCommand(String[] parameters, Snipe snipe) {
-        SnipeMessenger messenger = snipe.createMessenger();
-        if (parameters[1].equalsIgnoreCase("info")) {
-            messenger.sendMessage(ChatColor.GOLD + "Shell Voxel Parameters:");
-        } else {
-            messenger.sendMessage(ChatColor.RED + "Invalid parameter - see the info message for help.");
-        }
-    }
-
```
The info parameter is appended nevertheless.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stamp/CloneStampBrush.java:
```diff
@@ -37,7 +41,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             messenger.sendMessage(ChatColor.AQUA + "Default stamp brush");
         } else if (parameter.charAt(0) == 'c') {
-            Integer cylinderCenter = NumericParser.parseInteger(parameter.replace("c", ""));
+            Integer cylinderCenter = NumericParser.parseInteger(parameter.substring(1));
```
See FlatOceanBrush comment.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilBrush.java:
```diff
@@ -49,7 +51,7 @@
+        String firstParameter = parameters[0];
         byte pasteParam;
         if (firstParameter.equalsIgnoreCase("info")) {
             messenger.sendMessage(ChatColor.GOLD + "Stencil brush Parameters:");
```
One line below, `b schem`. That arg is invalid, the brush doesn't register itself under that name.
Actually, looking at the way args are sliced, using the command without input `/b stencil replace` causes an AIOOBE.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilListBrush.java:
```diff
@@ -37,7 +39,7 @@
+        String secondParameter = parameters[0];
         if (secondParameter.equalsIgnoreCase("info")) {
             messenger.sendMessage(ChatColor.GOLD + "Stencil List brush Parameters:");
             messenger.sendMessage(ChatColor.AQUA + "/b schem [optional: 'full' 'fill' or 'replace', with fill as default] [name] -- Loads the specified stencil list.  Full/fill/replace must come first.  Full = paste all blocks, fill = paste only into air blocks, replace = paste full blocks in only, but replace anything in their way.");
```
Same here, schem is invalid.

Aurelien30000 committed:
2.2.0

Co-authored-by: NotMyFault <mc.cache@web.de>

NotMyFault committed:
Improve gh artifact workflow

Aurelien30000 committed:
"Tree" brushes will be fixed in a future FAWE version
- FAVS workarounds are no longer needed.

Aurelien30000 committed:
Rework and improve brush commands:
- Fix a few technical issues
- Fix lots of inconsistencies
- Refactor EntityRemovalBrush

Aurelien30000 committed:
Add missing space

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/Brush.java:
```diff
@@ -25,12 +44,17 @@
 
     /**
-     * The powder action. Executed when a player RightClicks with Gunpowder
+     * The gunpowder action. Executed when a player RightClicks with Gunpowder
```
I will outline this here, (our) javadocs are written in English. There's no need to have random words capitalized unless they are reflecting to game internals or other cases where capitalization makes sense. But neither "Gunpowder", nor "RightClicks" are appropriate terms to describe the actions. Those can be changed to "right clicks" and "gunpowder".
Further occurrences will be referenced with "Spelling".

Aurelien30000 replied:
Ty, also changed ``Arrow`` to ``arrow``.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/AbstractBrush.java:
```diff
@@ -37,6 +44,36 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+     */
+    public List<String> sortCompletions(Stream<String> completions, String parameter, int index) {
+        // The first brush parameter may be info.
+        // Removing MINECRAFT_IDENTIFIER permits to complete whether minecraft:XXXX or XXXX.
```
completing > to completes (imo)

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/BiomeBrush.java:
```diff
@@ -2,52 +2,73 @@
+                if (firstParameter.equalsIgnoreCase("list")) {
+                    messenger.sendMessage(
+                            BiomeTypes.values().stream()
+                                    .map(biomeType -> ((biomeType == this.biomeType) ? ChatColor.GRAY : ChatColor.DARK_GRAY) +
```
![](https://i.imgur.com/xMgtUvu.png)
Why do I have one biome color out of place? Also, preferable they can be in light gray. I find dark colors always hard to read ingame.

Aurelien30000 replied:
Highlighted type is the selected one.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/CometBrush.java:
```diff
@@ -12,35 +12,59 @@
-                    messenger.sendMessage("Your balls are " + ChatColor.DARK_RED + ("small"));
+        String firstParameter = parameters[0];
+
+        if (firstParameter.equalsIgnoreCase("info")) {
```
Putting this here, if you set the comet brush, we could append a note that this brush is not undo-able due to the way it works but you can use `//ex` to extinguish the fire it creates.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/GenerateTreeBrush.java:
```diff
@@ -178,12 +93,165 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+                    BlockType woodType = BlockTypes.get(parameters[1]);
+                    if (woodType != null) {
+                        this.woodType = woodType;
+                        messenger.sendMessage(ChatColor.BLUE + "Wood Type set to " + this.woodType.getRichName().toString());
```
![](https://i.imgur.com/aTAZJpl.png)

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/GenerateTreeBrush.java:
```diff
@@ -178,12 +93,165 @@ public void handleCommand(String[] parameters, Snipe snipe) {
+                    BlockType leafType = BlockTypes.get(parameters[1]);
+                    if (leafType != null) {
+                        this.leafType = leafType;
+                        messenger.sendMessage(ChatColor.BLUE + "Leaf Type set to " + this.leafType.getRichName().toString());
```
See comment on wood type, same applies here

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/PullBrush.java:
```diff
@@ -23,15 +23,26 @@
+        SnipeMessenger messenger = snipe.createMessenger();
+        String firstParameter = parameters[0];
+        if (firstParameter.equalsIgnoreCase("info")) {
+            messenger.sendMessage(ChatColor.GOLD + "Pull Brush Parameters:");
```
How does one use this brush? `/b pull d 60` does nothing

Aurelien30000 replied:
Just use ``/b pull 60``

NotMyFault replied:
Oic, that works. It just doesn't respond anything. Also I think we should change letters that aren't `n` representing numbers. For sake of consistency and making it clearer that it's a number and not a parameter, `d` is often used as depth parameter.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RulerBrush.java:
```diff
@@ -20,35 +24,55 @@
+            messenger.sendMessage(ChatColor.GOLD + "Ruler Brush instructions: Right click first point with the arrow. " +
+                    "Right click with gunpowder for distances from that block (can repeat without getting a new first " +
+                    "block.) For placing blocks, use arrow and input the desired coordinates with parameters.");
+            messenger.sendMessage(ChatColor.LIGHT_PURPLE + "/b r [x] [y] [z] -- Places blocks one at a time of the type you " +
```
Using 2 args will cause an AIOOBE. Using 3 args in the following format `/b r 50 50 50` complaints about the format.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -44,85 +42,132 @@ public SignOverwriteBrush(File pluginDataFolder) {
+                    "internal buffer.");
+            messenger.sendMessage(ChatColor.DARK_AQUA + "Colors can be used using the \"&\" symbol.");
+            messenger.sendMessage(ChatColor.GOLD + "Sign Overwrite Brush Parameters:");
+            messenger.sendMessage(ChatColor.AQUA + "/b sio 1 [set|toggle] (...) -- Sets the text of the first sign line. " +
```
/b sio 1 set 50 returns nothing but works. This also affects the other args

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -44,85 +42,132 @@ public SignOverwriteBrush(File pluginDataFolder) {
-                        continue;
-                    }
-                    String fileName = parameters[++index];
+                } else if (firstParameter.equalsIgnoreCase("open") || firstParameter.equalsIgnoreCase("o")) {
```
Attempting to load an invalid file sets the default?
![](https://i.imgur.com/1c15y4L.png)

Aurelien30000 replied:
Buffer has always those default values, however shouldn't be sent if file loading fails, ty.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SignOverwriteBrush.java:
```diff
@@ -18,20 +19,17 @@
+import java.util.List;
+import java.util.stream.Stream;
 
 public class SignOverwriteBrush extends AbstractBrush {
```
Putting this here, but SignOverWriteBrush#saveBufferToFile should save them in a dedicated folder, like `/signoverwrite/`. Else this gets too messy imo.
Same behavior can be found in stencil files. This is the only one that has no dedicated folder.

Aurelien30000 replied:
hell yes!

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SpiralStaircaseBrush.java:
```diff
@@ -12,6 +12,7 @@
+import java.util.List;
 import java.util.stream.Stream;
 
 public class SpiralStaircaseBrush extends AbstractBrush {
```
How is this brush supposed to work? And why can I only use two types of stair cases?
/b sstair woodstair
![](https://i.imgur.com/AOhQVFk.png)

Aurelien30000 replied:
No idea, just ported the original voxel brush 🤷

NotMyFault replied:
I found this description at some wiki:

> You can use this to create a spiral stair case for you. The arrow will remove any surrounding blocks within the radius you specified and spawn the stair case. You can use the gunpowder tool for this however that tool will not let you go below grass level.

NotMyFault replied:
Note: Just tested on 1.12 with legacy VS and the outcome is the same. So not something that is newly "broken", dunno if you can call this broken, idk what it's supposed to do, but to my understanding it should create an actual spiral staircase around the targeted block?

Aurelien30000 replied:
You must set a voxel height (``/vh [n]``), will add a note, tried with 10, got this build:

![2021-08-12_23 26 55](https://user-images.githubusercontent.com/43724816/129272339-2f9be88e-f486-4515-9d2d-ee988a02469d.png)

NotMyFault replied:
Good to know, maybe add an arg to pass the width of the staircase? Or is it always one block 👀

Aurelien30000 replied:
Probably needs a rewrite to accept all stair types, another pr later tbh, sorry.

NotMyFault replied:
Nw, bad enough it's hard coded within 30+ occurrences in the SpiralStairCase class

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/TreeSnipeBrush.java:
```diff
@@ -10,36 +10,66 @@
 import java.util.stream.IntStream;
+import java.util.stream.Stream;
 
 public class TreeSnipeBrush extends AbstractBrush {
```
See my comment on generate tree, it may apply here too
![](https://i.imgur.com/pbSRSfA.png)

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/canyon/CanyonBrush.java:
```diff
@@ -7,8 +7,12 @@
+import java.util.List;
+import java.util.stream.Stream;
+
 public class CanyonBrush extends AbstractBrush {
```
I'm unsure whether this should be addressed here or in Fawe, however, it doesn't calculate light properly
![](https://i.imgur.com/1aCyqQU.jpg)

Aurelien30000 replied:
_Huphm, no idea again, light issues_ 😭
Works fine on Tuinity (except on chunk edges, whereas it should), seems FAWE implementation related 🤷 
![2021-08-12_23 54 48](https://user-images.githubusercontent.com/43724816/129275118-4d57c5ee-3d00-4965-a042-6a1902eaed8b.png)

NotMyFault replied:
Meh, I'll ignore this for now. We have https://github.com/IntellectualSites/FastAsyncWorldEdit-Adapters/issues/6 in place.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/PunishBrush.java:
```diff
@@ -41,56 +50,88 @@
+                if (firstParameter.equalsIgnoreCase("list")) {
+                    messenger.sendMessage(
+                            Arrays.stream(Punishment.values())
+                                    .map(punishment -> ((punishment == this.punishment) ? ChatColor.GRAY : ChatColor.DARK_GRAY) +
```
See previous color related comments 
![](https://i.imgur.com/YLalLEi.png)

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation2DBrush.java:
```diff
@@ -20,9 +21,26 @@
+
+        if (firstParameter.equalsIgnoreCase("info")) {
+            messenger.sendMessage(ChatColor.GOLD + "Rotation2D Brush Parameters:");
+            messenger.sendMessage(ChatColor.AQUA + "/b rot2 [a] -- Sets rotation angle to a.");
```
That should be n
How does it parse inputs?
![](https://i.imgur.com/BXkemFT.png)

Aurelien30000 replied:
Angle is converted to gradians, I'll add a note after the value.

NotMyFault replied:
Putting this here, if you find some other args that aren't `n` but represent a number (therefore aren't a needed parameter), please replace it for sake of consistency. Having too many aliases for representations in place is just confusing.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation2DVerticalBrush.java:
```diff
@@ -24,13 +24,25 @@
+
+        if (firstParameter.equalsIgnoreCase("info")) {
+            messenger.sendMessage(ChatColor.GOLD + "Rotation2DVertical Brush Parameters:");
+            messenger.sendMessage(ChatColor.AQUA + "/b rot2v [a] -- Sets rotation angle to a");
```
See Rotation2dBrush comment

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/rotation/Rotation3DBrush.java:
```diff
@@ -25,39 +29,50 @@
+        // which way is clockwise is less obvious for roll and pitch... should probably fix that / make it clear
+        if (firstParameter.equalsIgnoreCase("info")) {
+            messenger.sendMessage(ChatColor.GOLD + "Rotate Brush Brush Parameters:");
+            messenger.sendMessage(ChatColor.AQUA + "/b rot3d p [a] -- Sets degrees of pitch rotation to a (rotation about the Z" +
```
`/b info rot3` returns with rot3d args, but you can't use them. It identifies itself as rot3

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilBrush.java:
```diff
@@ -49,39 +51,60 @@
     @Override
```
Is this still somewhat of used?
![](https://i.imgur.com/rvFaEky.png)
Seems like a weird way to restrict a command. Nor have I seen it somewhere else yet.

Aurelien30000 replied:
Not at all, let's remove this note.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilBrush.java:
```diff
@@ -49,39 +51,60 @@
+            messenger.sendMessage(ChatColor.DARK_AQUA + "Size of the stencils you are allowed to paste depends on rank (member / " +
+                    "lite, sniper, curator, admin)");
+            messenger.sendMessage(ChatColor.GOLD + "Stencil Brush Parameters:");
+            messenger.sendMessage(ChatColor.AQUA + "/b st (full|fill|replace) [n] -- Loads the specified stencil n. Allowed " +
```
Is n supposed to be a number? If not change it. I'd say that n is reserved for any numbers in any place.

Aurelien30000 replied:
S like String is more likely appropriate.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilListBrush.java:
```diff
@@ -32,39 +34,63 @@
+        String firstParameter = parameters[0];
+
+        if (firstParameter.equalsIgnoreCase("info")) {
+            messenger.sendMessage(ChatColor.GOLD + "Stencil List Brush Parameters:");
```
See former comment

NotMyFault commented on .github/workflows/build.yml:
```diff
@@ -8,10 +8,30 @@ jobs:
       - name: "Setup JDK 16"
         uses: "actions/setup-java@v2.1.0"
         with:
           distribution: "adopt"
```
2.2.0 & temurin. AOJDK moved to the Eclipse Foundation, therefore newer versions are deployed to Adoptium instead of AdoptOpenJDK.

Aurelien30000 committed:
Fix some issues, address most requests

NotMyFault committed:
Add `info` command

NotMyFault committed:
Merge branch 'main' of https://github.com/Aurelien30000/FastAsyncVoxelSniper into pr/50

NotMyFault committed:
Update maven version

Aurelien30000 reviewed (commented):
👍 Should be good overall.
